File: src/commands/note.ts
import type { Command } from "commander";
import { loadConfig } from "../config.ts";
import { ensureStoreForWrite } from "../data.ts";
import { printJSON } from "../envelope.ts";
import { isValidYMD } from "../models.ts";
import {
  filterNotes,
  localISO,
  NOTE_AUTHORS,
  NOTE_TYPES,
  type NoteAuthor,
  type NoteRecord,
  type NoteType,
  readAllNotes,
  ulid,
  writeNote,
} from "../notes.ts";
import { dataPaths } from "../paths.ts";
import { readWorkouts } from "../storage.ts";
import { resolveWorkout } from "./show.ts";

async function readBody(bodyArg: string | undefined): Promise<string> {
  if (bodyArg != null && bodyArg !== "-") return bodyArg.trim();
  const text = await new Response(Bun.stdin.stream()).text();
  return text.trim();
}

export function parseNoteDate(raw: string): string | null {
  const prefix = /^(\d{4}-\d{2}-\d{2})/.exec(raw);
  if (prefix != null && !isValidYMD(prefix[1]!)) return null;
  const d = isValidYMD(raw) ? new Date(`${raw}T12:00:00`) : new Date(raw);
  if (Number.isNaN(d.getTime())) return null;
  return localISO(d);
}

function noteLine(n: NoteRecord): string {
  const workout = n.workout_id != null ? ` w:${n.workout_id}` : "";
  const tags = n.tags && n.tags.length > 0 ? ` #${n.tags.join(" #")}` : "";
  return `${n.date.slice(0, 10)}  [${n.type}/${n.author}]${workout}${tags}  ${n.body}`;
}

export function registerNote(program: Command): void {
  const note = program.command("note").description("Coaching notes and subjective reports");

  note
    .command("add [body]")
    .description("Record a note (body as argument, or '-' / omitted to read stdin)")
    .option("--type <type>", "subjective, observation, or lesson", "observation")
    .option("--workout <id>", "link to a workout id or 'last'")
    .option("--tags <tags>", "comma-separated tags")
    .option("--author <author>", "athlete or coach", "athlete")
    .option("--date <date>", "backdate the note (YYYY-MM-DD or ISO timestamp)")
    .action(
      async (
        bodyArg: string | undefined,
        opts: { type: string; workout?: string; tags?: string; author: string; date?: string },
      ) => {
        if (!(NOTE_TYPES as readonly string[]).includes(opts.type)) {
          console.error(`Error: --type must be one of ${NOTE_TYPES.join(", ")}.`);
          process.exit(1);
        }
        if (!(NOTE_AUTHORS as readonly string[]).includes(opts.author)) {
          console.error(`Error: --author must be one of ${NOTE_AUTHORS.join(", ")}.`);
          process.exit(1);
        }

        const body = await readBody(bodyArg);
        if (body === "") {
          console.error("Error: note body is empty.");
          process.exit(1);
        }

        const cfg = await loadConfig();
        const paths = dataPaths(cfg);
        const now = new Date();
        const storeError = await ensureStoreForWrite(paths, now);
        if (storeError != null) {
          console.error(storeError);
          process.exit(1);
        }

        let workoutId: number | undefined;
        if (opts.workout) {
          const workouts = await readWorkouts(paths);
          const w = resolveWorkout(workouts, opts.workout);
          if (w == null) {
            console.error(`Error: no workout matching "${opts.workout}".`);
            process.exit(1);
          }
          workoutId = w.id;
        }

        let date = localISO(now);
        if (opts.date) {
          const parsed = parseNoteDate(opts.date);
          if (parsed == null) {
            console.error(`Error: invalid --date "${opts.date}".`);
            process.exit(1);
          }
          date = parsed;
        }

        const record: NoteRecord = {
          id: ulid(now),
          date,
          type: opts.type as NoteType,
          workout_id: workoutId,
          tags: opts.tags
            ? opts.tags
                .split(",")
                .map((t) => t.trim())
                .filter((t) => t !== "")
            : undefined,
          body,
          author: opts.author as NoteAuthor,
        };
        await writeNote(paths, record);
        console.log(record.id);
      },
    );

  note
    .command("list")
    .description("List notes, newest last")
    .option("--type <type>", "filter by type")
    .option("--since <date>", "only notes on or after date (YYYY-MM-DD)")
    .option("--workout <id>", "only notes linked to a workout id")
    .option("-n, --count <n>", "show only the most recent n notes")
    .option("--json", "output as JSON")
    .action(
      async (opts: {
        type?: string;
        since?: string;
        workout?: string;
        count?: string;
        json?: boolean;
      }) => {
        if (opts.since && !isValidYMD(opts.since)) {
          console.error(`Error: invalid --since date "${opts.since}" (expected YYYY-MM-DD).`);
          process.exit(1);
        }
        const cfg = await loadConfig();
        const paths = dataPaths(cfg);
        let notes = filterNotes(await readAllNotes(paths), {
          type: opts.type,
          since: opts.since,
          workoutId: opts.workout != null ? Number(opts.workout) : undefined,
        });
        if (opts.count != null) {
          const n = parseInt(opts.count, 10);
          if (Number.isNaN(n) || n < 1) {
            console.error("Error: --count must be a positive integer.");
            process.exit(1);
          }
          notes = notes.slice(-n);
        }

        if (opts.json) {
          printJSON("c2.notes.v1", { count: notes.length, notes });
          return;
        }
        if (notes.length === 0) {
          console.log("No notes found.");
          return;
        }
        for (const n of notes) {
          console.log(noteLine(n));
        }
      },
    );

  note
    .command("show <id>")
    .description("Show one note in full")
    .option("--json", "output as JSON")
    .action(async (id: string, opts: { json?: boolean }) => {
      const cfg = await loadConfig();
      const paths = dataPaths(cfg);
      const match = (await readAllNotes(paths)).find((n) => n.id === id);
      if (match == null) {
        console.error(`No note with id ${id}.`);
        process.exit(1);
      }
      if (opts.json) {
        printJSON("c2.note.v1", match);
        return;
      }
      console.log(`Id: ${match.id}`);
      console.log(`Date: ${match.date}`);
      console.log(`Type: ${match.type} (${match.author})`);
      if (match.workout_id != null) console.log(`Workout: ${match.workout_id}`);
      if (match.tags && match.tags.length > 0) console.log(`Tags: ${match.tags.join(", ")}`);
      console.log();
      console.log(match.body);
    });
}

File: internal/cli/note.go
package cli

import (
	"fmt"
	"io"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/richhaase/c2/internal/envelope"
	"github.com/richhaase/c2/internal/models"
	"github.com/richhaase/c2/internal/notes"
	"github.com/richhaase/c2/internal/storage"
	"github.com/richhaase/c2/internal/store"
)

type notesPayload struct {
	Count int            `json:"count"`
	Notes []notes.Record `json:"notes"`
}

var noteDatePattern = regexp.MustCompile(`^(\d{4}-\d{2}-\d{2})([T ].+)?$`)

var offsetSuffixPattern = regexp.MustCompile(`(?:[+-]\d{2}:\d{2}|Z)$`)

func parseNoteDate(raw string) (string, bool) {
	m := noteDatePattern.FindStringSubmatch(raw)
	if m == nil || !models.IsValidYMD(m[1]) {
		return "", false
	}
	if offsetSuffixPattern.MatchString(raw) {
		normalized := strings.Replace(raw, " ", "T", 1)
		if strings.HasSuffix(normalized, "Z") {
			normalized = normalized[:len(normalized)-1] + "+00:00"
		}
		if !regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\S+$`).MatchString(normalized) {
			return "", false
		}
		if _, ok := notes.ParseDate(normalized); !ok {
			return "", false
		}
		return normalized, true
	}
	if m[2] == "" {
		t := models.ParseLocal(raw + "T12:00:00")
		if t.IsZero() {
			return "", false
		}
		return notes.LocalISO(t), true
	}
	t := models.ParseLocal(raw)
	if t.IsZero() {
		return "", false
	}
	return notes.LocalISO(t), true
}

func readBody(cmd *cobra.Command, arg string) (string, error) {
	if arg != "" && arg != "-" {
		return strings.TrimSpace(arg), nil
	}
	data, err := io.ReadAll(cmd.InOrStdin())
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(data)), nil
}

func noteLine(n notes.Record) string {
	workout := ""
	if n.WorkoutID != nil {
		workout = fmt.Sprintf(" w:%d", *n.WorkoutID)
	}
	tags := ""
	if len(n.Tags) > 0 {
		tags = " #" + strings.Join(n.Tags, " #")
	}
	return fmt.Sprintf("%s  [%s/%s]%s%s  %s", n.Date[:10], n.Type, n.Author, workout, tags, n.Body)
}

func newNoteCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "note",
		Short: "Coaching notes and subjective reports",
	}
	cmd.AddCommand(newNoteAddCmd(), newNoteListCmd(), newNoteShowCmd())
	return cmd
}

func newNoteAddCmd() *cobra.Command {
	var (
		noteType string
		workout  string
		tags     string
		author   string
		dateFlag string
	)
	cmd := &cobra.Command{
		Use:   "add [body]",
		Short: "Record a note (body as argument, or '-' / omitted to read stdin)",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if !slices.Contains(notes.Types, noteType) {
				return reportf(cmd, "Error: --type must be one of %s.", strings.Join(notes.Types, ", "))
			}
			if !slices.Contains(notes.Authors, author) {
				return reportf(cmd, "Error: --author must be one of %s.", strings.Join(notes.Authors, ", "))
			}

			backdate := ""
			if dateFlag != "" {
				parsed, ok := parseNoteDate(dateFlag)
				if !ok {
					return reportf(cmd, "Error: invalid --date %q.", dateFlag)
				}
				backdate = parsed
			}

			bodyArg := ""
			if len(args) > 0 {
				bodyArg = args[0]
			}
			body, err := readBody(cmd, bodyArg)
			if err != nil {
				return err
			}
			if body == "" {
				return reportf(cmd, "Error: note body is empty.")
			}

			_, p, err := loadStore()
			if err != nil {
				return err
			}
			now := time.Now()
			warn := warner(cmd)
			if err := store.RejectForeign(p, warn); err != nil {
				return reportf(cmd, "%v", err)
			}

			var workoutID *int64
			if workout != "" {
				workouts, err := storage.ReadWorkouts(p)
				if err != nil {
					return err
				}
				w := models.ResolveWorkout(workouts, workout)
				if w == nil {
					return reportf(cmd, "Error: no workout matching %q.", workout)
				}
				workoutID = &w.ID
			}

			if err := store.EnsureForWrite(p, now, warn); err != nil {
				return reportf(cmd, "%v", err)
			}

			date := backdate
			if date == "" {
				date = notes.LocalISO(now)
			}

			var tagList []string
			if tags != "" {
				for _, t := range strings.Split(tags, ",") {
					if trimmed := strings.TrimSpace(t); trimmed != "" {
						tagList = append(tagList, trimmed)
					}
				}
			}

			record := notes.Record{
				ID:        notes.ULID(now),
				Date:      date,
				Type:      noteType,
				WorkoutID: workoutID,
				Tags:      tagList,
				Body:      body,
				Author:    author,
			}
			if !notes.IsShaped(record) {
				return reportf(cmd, "Error: note would not round-trip; check --date and field values.")
			}
			if err := notes.Write(p, record); err != nil {
				return err
			}
			fmt.Fprintln(cmd.OutOrStdout(), record.ID)
			return nil
		},
	}
	cmd.Flags().StringVar(&noteType, "type", "observation", "subjective, observation, or lesson")
	cmd.Flags().StringVar(&workout, "workout", "", "link to a workout id or 'last'")
	cmd.Flags().StringVar(&tags, "tags", "", "comma-separated tags")
	cmd.Flags().StringVar(&author, "author", "athlete", "athlete or coach")
	cmd.Flags().StringVar(&dateFlag, "date", "", "backdate the note (YYYY-MM-DD or ISO timestamp)")
	return cmd
}

func newNoteListCmd() *cobra.Command {
	var (
		noteType string
		since    string
		workout  string
		count    string
		asJSON   bool
	)
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List notes, newest last",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			if since != "" && !models.IsValidYMD(since) {
				return reportf(cmd, "Error: invalid --since date %q (expected YYYY-MM-DD).", since)
			}
			if noteType != "" && !slices.Contains(notes.Types, noteType) {
				return reportf(cmd, "Error: --type must be one of %s.", strings.Join(notes.Types, ", "))
			}
			var workoutID *int64
			if workout != "" {
				id, err := strconv.ParseInt(workout, 10, 64)
				if err != nil {
					return reportf(cmd, "Error: invalid --workout id %q.", workout)
				}
				workoutID = &id
			}

			_, p, err := loadStore()
			if err != nil {
				return err
			}
			if err := store.RejectForeign(p, warner(cmd)); err != nil {
				return reportf(cmd, "%v", err)
			}
			matched := notes.Apply(notes.ReadAll(p), notes.Filter{
				Type:      noteType,
				Since:     since,
				WorkoutID: workoutID,
			})
			if count != "" {
				n, err := positiveInt(cmd, count, "Error: --count must be a positive integer.")
				if err != nil {
					return err
				}
				if n < len(matched) {
					matched = matched[len(matched)-n:]
				}
			}

			out := cmd.OutOrStdout()
			if asJSON {
				if matched == nil {
					matched = []notes.Record{}
				}
				return envelope.Print(out, "c2.notes.v1", notesPayload{Count: len(matched), Notes: matched})
			}
			if len(matched) == 0 {
				fmt.Fprintln(out, "No notes found.")
				return nil
			}
			for _, n := range matched {
				fmt.Fprintln(out, noteLine(n))
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&noteType, "type", "", "filter by type")
	cmd.Flags().StringVar(&since, "since", "", "only notes on or after date (YYYY-MM-DD)")
	cmd.Flags().StringVar(&workout, "workout", "", "only notes linked to a workout id")
	cmd.Flags().StringVarP(&count, "count", "n", "", "show only the most recent n notes")
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}

func newNoteShowCmd() *cobra.Command {
	var asJSON bool
	cmd := &cobra.Command{
		Use:   "show <id>",
		Short: "Show one note in full",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			_, p, err := loadStore()
			if err != nil {
				return err
			}
			if err := store.RejectForeign(p, warner(cmd)); err != nil {
				return reportf(cmd, "%v", err)
			}
			var match *notes.Record
			for _, n := range notes.ReadAll(p) {
				if n.ID == args[0] {
					match = &n
					break
				}
			}
			if match == nil {
				return reportf(cmd, "No note with id %s.", args[0])
			}

			out := cmd.OutOrStdout()
			if asJSON {
				return envelope.Print(out, "c2.note.v1", match)
			}
			fmt.Fprintf(out, "Id: %s\n", match.ID)
			fmt.Fprintf(out, "Date: %s\n", match.Date)
			fmt.Fprintf(out, "Type: %s (%s)\n", match.Type, match.Author)
			if match.WorkoutID != nil {
				fmt.Fprintf(out, "Workout: %d\n", *match.WorkoutID)
			}
			if len(match.Tags) > 0 {
				fmt.Fprintf(out, "Tags: %s\n", strings.Join(match.Tags, ", "))
			}
			fmt.Fprintln(out)
			fmt.Fprintln(out, match.Body)
			return nil
		},
	}
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}
